File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerULong.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) ulong values to (and from) a binary data source.
/// </summary>
public class BinarySerializerULong : IBinaryTypeSerializer<ulong>
{
    #region Constants
        
    private const int ENCODED_LENGTH_IN_BYTES = sizeof(ulong);
        
    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, ulong value)
    {
        writer.WriteBytes(BitConverter.GetBytes(value));
    }

    /// <inheritdoc />
    public ulong Decode(IDataBufferReader reader)
    {
        return BitConverter.ToUInt64(reader.ReadBytes(ENCODED_LENGTH_IN_BYTES));
    }
        
    /// <inheritdoc />
    public bool GetEquals(ulong a, ulong b)
    {
        return a.Equals(b);
    }

    #endregion
}

File: typescript/src/types/binary-serializer-ulong.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";

/**
 * Reads (and writes) ulong values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<bigint>}
 */
export class BinarySerializerULong implements BinaryTypeSerializer<bigint> {
    get sizeInBytes(): number {
        return 8;
    }

    encode(writer: DataWriter, value: bigint): void {
        const buffer = new ArrayBuffer(this.sizeInBytes);
        const view = new DataView(buffer);
        view.setBigUint64(0, value, true);
        
        writer.writeBytes(new Uint8Array(buffer));
    }

    decode(reader: DataReader): bigint {
        const valueBytes = reader.readBytes(this.sizeInBytes);

        return new DataView(valueBytes.buffer, valueBytes.byteOffset, valueBytes.byteLength).getBigUint64(0, true);
    }

    getEquals(a: bigint, b: bigint): boolean {
        return a === b;
    }
}